pxteditor/code-validation/runValidatorPlanAsync.ts

```diff
@@ -42,13 +57,11 @@ export async function runValidatorPlanAsync(usedBlocks: Blockly.Block[], plan: p
     return passed;
 }
 
-function runBlocksExistValidation(usedBlocks: Blockly.Block[], inputs: pxt.blocks.BlocksExistValidatorCheck): boolean {
+function runBlocksExistValidation(usedBlocks: Blockly.Block[], inputs: pxt.blocks.BlocksExistValidatorCheck): [Blockly.Block[], boolean] {
     const blockResults = validateBlocksExist({ usedBlocks, requiredBlockCounts: inputs.blockCounts });
-    const success =
-        blockResults.disabledBlocks.length === 0 &&
-        blockResults.missingBlocks.length === 0 &&
-        blockResults.insufficientBlocks.length === 0;
-    return success;
+    const blockId = Object.keys(inputs.blockCounts)[0];
```

Comment:
This method has several array accesses without range checks. Are these arrays guaranteed to be non-empty?

Reply:
For `blockId` in particular, if `Object.keys(inputs.blockCounts)` returned nothing, then the validation would be essentially pointless. With this array access, I'm getting the block that we're checking exists for that validator run. 

So, for example, if you have this validator:

>         {
>             ".desc": "A custom variable's value is set to a random number",
>             "name": "variable_set_random",
>             "threshold": 1,
>             "checks": [
>                 {
>                     "validator": "blocksExist",
>                     "blockCounts": {
>                         "variables_set": 1
>                     },
>                     "childValidatorPlans": ["device_random_used"]
>                 }
>             ]
>         }

The block that we're actually wanting to check for is `variables_set`, and you can only know this in this function by looking in the JSON. For tutorial validation, you can have multiple different blocks in the `blockCounts` object, but for our validation, we should only ever have one entry in the `blockCounts` object. Either way, if  `Object.keys(inputs.blockCounts)` was empty at the point that we do an array access in this function, validation would be essentially pointless. However, you're right in that I should check whether the array returned from `Object.keys()` has length before accessing.